cli/src/lib.rs

```diff
@@ -3022,6 +3035,7 @@ fn run_test_suite(
 fn validator_flags(
     cfg: &WithPath<Config>,
     test_validator: &Option<TestValidator>,
+    test_upgradeable_program: bool,
```

Comment:
This looks good to me but it would be better if you could add the `--upgradeable-program` argument to the `Validator` struct here https://github.com/coral-xyz/anchor/blob/d1e32674d50005d04160b6fbac6f91de602a8829/cli/src/config.rs#L1030 and remove the extra parameter `test_upgradeable_program` from the `validator_flags` function. That way we'll both have access to the `--upgradeable-program` argument(via `test_validator` argument) which could be used in other places and also remove the extra argument from the function.

Reply:
Ahh very cool!  Instead of an `anchor test` CLI option, would we enable the feature via `anchor.toml` like this? 
```
[test.validator]
upgradeable_program = true
```

I've made the code changes.  PTAL again as I initially had made a mistake:  `--upgradeable-program` requires 3 params with the third being the upgrade authority's pubkey.

Reply:
I think we can keep the `--upgradeable-program`(or `--upgradeable`) flag for the `anchor test` command for ease of use(like how it was in the initial PR) but also follow the convention of the `solana-test-validator` arguments for the `[test.validator]` entries.

For example `--account` argument is stored as `Option<Vec<AccountEntry>>` https://github.com/coral-xyz/anchor/blob/d1e32674d50005d04160b6fbac6f91de602a8829/cli/src/config.rs#L1030-L1032

Similarly we could store the `--upgradeable-program` as `Option<Vec<UpgradeableProgramEntry>>` to keep the convention the same(instead of storing `Option<bool>`).

Reply:
It looks like the "primary" programs to test are stored separately from the `test.validator` and `test.genesis` entries. https://github.com/nickguo/anchor/blob/2d00b489b44ce045cbc39468d38a4cf7e94dd823/cli/src/lib.rs#L3039 

This is making it a bit tricky as the new `Option<Vec<UpgradeableProgramEntry>>` doesn't have an entry for the main program that we want to test.  So for the main program, we'd have to use something like the initial `    test_upgradeable_program: bool` param to `validator_flags()` which wasn't ideal.

Overall it looks a bit of a code smell -- it should be possible to remove `test.genesis` altogether, and have all programs-to-load configured under `test.validator`.  The main programs can populated there by default.  On top of `UpgradeableProgramEntry` would also add `Option<Vec<BpfProgramEntry>>` which has been missing (functionally the same as the current `test.genesis`).

Sorry for the long post!  I see two ways forward:
(1) Configure upgradeable main program differently from "genesis" programs (e.g. pyth).  For upgradeable main program, we would pass a CLI flag to `anchor test`.  For genesis programs, we have that optional bool.

(2) Refactor away `test.genesis` and instead populate `test.validator`.  We can do this in a backwards compatible way.  And then main program and additional genesis programs are loaded the same way.  Finally, we can add the upgradeable functionality afterwards.

What do you think?

Reply:
> (1) Configure upgradeable main program differently from "genesis" programs (e.g. pyth). For upgradeable main program, we would pass a CLI flag to `anchor test`. For genesis programs, we have that optional bool.

If I understand correctly, option 1 is roughly the same as [my last post](https://github.com/coral-xyz/anchor/pull/2642#discussion_r1341949281) but both options seem viable.

Reply:
Went with (1) !  

I decided to bubble up the flag from `Validator` into `TestValidator` because the tricky thing was that `Validator` does not have entries for the programs to deploy.  PTAL 🙏

Reply:
And went with `anchor.toml` to configure whether to use upgradeable program.  The CLI flag felt clunky in comparison -- both the ergonomics to use it and the code changes to add it.

Updated `docs/src/pages/docs/manifest.md` to add the new setting

Reply:
Awesome, thank you for updating the docs as well!